include/pybind11/stl.h

```diff
@@ -305,6 +314,55 @@ template<> struct type_caster<std::experimental::nullopt_t>
     : public void_caster<std::experimental::nullopt_t> {};
 #endif
 
+#if defined(PYBIND11_HAS_FILESYSTEM)
+template<> struct type_caster<std::filesystem::path> {
+
+    static handle cast(const std::filesystem::path& path, return_value_policy, handle) {
+        PyObject* py_str = [](const auto& w) {
+            using char_type = std::filesystem::path::value_type;
+            if constexpr (std::is_same_v<char_type, char>) {
+                return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
+            } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+                return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
+            }
+        } (path.native());
+        if (!py_str) {
+            return nullptr;
+        }
+        return module::import("pathlib").attr("Path")(reinterpret_steal<object>(py_str))
+               .release();
+    }
+
+    bool load(handle handle, bool) {
+        using char_type = std::filesystem::path::value_type;
+        PyObject* tmp = nullptr;
+        if constexpr (std::is_same_v<char_type, char>) {
+            if (PyUnicode_FSConverter(handle.ptr(), &tmp)) {
+                if (auto c_str = PyBytes_AsString(tmp)) {
+                    value = c_str;
+                }
+            }
+        } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+            if (PyUnicode_FSDecoder(handle.ptr(), &tmp)) {
+                if (auto c_str = PyUnicode_AsWideCharString(tmp, nullptr)) {
+                    value = c_str;
+                    PyMem_Free(c_str);
```

Comment:
Why is this `PyMem_Free` here? Looks to me like it's leaving `value` in an invalid state.

Reply:
AFAICT the assignment will *copy* the contents of the string; this can be checked by inserting e.g. `c_str[0] = 'z';` immediately after `value = c_str;`.  (I have checked that this is indeed the case on Linux; I haven't checked on Windows but that seems most likely to be the case too...)

In any case we need to call PyMem_Free somewhere to avoid a memleak, and it's certainly not the std::filesystem::path destructor that will do that for us...

Reply:
Aha! That ends up calling `value.operator=(c_str)` which will create a temporary `std::wstring`. That makes sense.

For anyone else wondering "Doesn't this mean that we need to free `c_str` in the other branch", no. The other branch with `char*` doesn't allocate.

Reply:
If this tripped @bstaletic, it might need a comment?

Reply:
sure, done.

Reply:
Thanks! :-)